### PR TITLE
(XL) Pageserve persist login session

### PR DESCRIPTION
**Pull request by mukobi**
Save the user's login information in secure `session` cookies via the pageserve service.

Complete test coverage for the routes and functions I modified/added.

### CODE REVIEW

**Comment by mukobi** [comment]
Sorry for all the small commits. It's probably easier to look through the files changed.

File: pageserve/templates/base.html [hunk 1]
```diff
@@ -20,39 +20,55 @@
...
+        }
+    </script>
 </head>
 
 <body>
     <header>
         {% block header %}{% endblock header %}
     </header>
     {% block login_buttons %}
-    {% if not signed_in %}
```
Comment (cmei4444) [inline pageserve/templates/base.html]: Where is signed_in passed to the template? Or does it automatically have the info?

File: pageserve/test_routes.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                    self.assert_sessions_are_equal(
+                        result_dict, VALID_USER_IN_SESSION)
+                    # user stored in session
+                    self.assertEqual(len(flask.session), 1)
+                    self.assertEqual(
+                        flask.session["user"], VALID_USER_IN_SESSION)
+
+    def test_failed_authentication(self):
+        """GAuth token fails authentication."""
+        response_content = "Error: bad gauth_token."
```
Comment (evankanderson) [inline pageserve/test_routes.py]: This style seems different from the style in the test above (i.e. define some constants before starting the context managers vs literals or global constants). Make them consistent (either style is fine).

File: pageserve/test_routes.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    def setUp(self):
+        """Set up test client."""
+        app.config["TESTING"] = True  # propagate exceptions to test client
+        app.secret_key = "Dummy key used for testing the flask session"
+
+    def test_valid_authentication(self):
+        """GAuth token successfully authenticates."""
+        with app.test_request_context():
+            with app.test_client() as client:
+                with mock.patch("app.requests") as mock_requests:
```
[the remark this answers is not in the document]
Comment (cmei4444) [reply pageserve/test_routes.py]: I actually don't think I've used it -  I might have looked at it but I don't see it anywhere in my code.

File: pageserve/app.py [hunk 4]
```diff
@@ -185,6 +242,7 @@ def config_endpoints(endpoints):
 app.config["GAUTH_CLIENT_ID"] = os.environ.get("GAUTH_CLIENT_ID")
 app.config["GAUTH_CALLBACK_ENDPOINT"] = (app.config['USERS_ENDPOINT']
                                          + "authenticate")
+app.secret_key = os.environ.get("FLASK_SECRET_KEY")
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/app.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/app.py [hunk 5]
```diff
@@ -15,47 +15,97 @@
...
 
 
 @app.route('/v1/events', methods=['GET'])
 def show_events():
     """Displays page with all sub-events."""
-    user = get_user()
-    is_auth = has_edit_access(get_user_info(user,
-                                            app.config['USERS_ENDPOINT']
-                                            + "authorization"))
+    is_auth = has_edit_access(get_user())
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/app.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/app.py [hunk 6]
```diff
@@ -151,22 +201,29 @@ def parse_events(events_dict):
...
-    return user_data['edit_access']
+    if user is None:
+        return False
+    url = app.config["USERS_ENDPOINT"] + "authorization"
+    response = requests.post(url, data={"user_id": user["user_id"]})
+    return bool(response.json()["edit_access"]) if user else False
 
 
 def get_user():
     """Retrieves the current user of the app."""
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/app.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/templates/base.html [hunk 7]
```diff
@@ -20,39 +20,55 @@
...
+        <div class="g-signin2" data-onsuccess="onSignIn" data-theme="dark"></div>
+        <script>
+            function onSignIn(googleUser) {
+                // Secure Google ID token to send to backend for verification.
+                var id_token = googleUser.getAuthResponse().id_token;
+                var xhr = new XMLHttpRequest();
+                xhr.open('POST', "authenticate");
+                xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
+                xhr.onload = function () {
+                    console.log('Signed in as: ' + xhr.responseText);
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/templates/base.html]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/templates/base.html [hunk 8]
```diff
@@ -20,39 +20,55 @@
...
-            });
-        }
-    </script>
+    <div id="sign_out_button">
+        <a href="#" onclick="signOut();">Sign out</a>
+        <script>
+            function signOut() {
+                var auth2 = gapi.auth2.getAuthInstance();
+                auth2.signOut().then(function () {
+                    console.log('User signed out.');
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/templates/base.html]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/test_routes.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+    def setUp(self):
+        """Set up test client."""
+        app.config["TESTING"] = True  # propagate exceptions to test client
+        app.secret_key = "Dummy key used for testing the flask session"
+
+    def test_valid_authentication(self):
+        """GAuth token successfully authenticates."""
+        with app.test_request_context():
+            with app.test_client() as client:
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/test_routes.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/test_routes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/test_routes.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/test_routes.py (hunk 2 above)
Comment (mukobi) [reply pageserve/test_routes.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/a74b1c700a7671fe772b22e0c8ed31a1d262a0ad

File: pageserve/app.py [hunk 10]
```diff
@@ -15,47 +15,97 @@
...
+        gauth_token: Google ID token to authenticate.
+
+    Response:
+        Response:
+            201: user object from the db if authentication was successful.
+            400: error message if authentication was not successful.
+    """
+    try:
+        gauth_token = request.form["gauth_token"]
+        response = authenticate_with_users_service(gauth_token)
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/app.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/b7946c9c3ebdfc5b05c71e05ab1ae72ba8a2be0f

File: pageserve/app.py [hunk 11]
```diff
@@ -15,47 +15,97 @@
...
+        response: response from the users service
+    """
+    response = requests.post(
+        app.config["USERS_ENDPOINT"] + "authenticate",
+        data={"gauth_token": gauth_token})
+    if response.status_code == 201:
+        # authentication successful, store login in cookie
+        session["user"] = response.json()
+        # also store the valid login token
+        session["user"]["gauth_token"] = gauth_token
```
[the remark this answers is not in the document]
Comment (mukobi) [reply pageserve/app.py]: Done https://github.com/knative-portability/large-events/pull/37/commits/4fb20f3fda0d3d58004e99f345f92773f62534dd. I chose 3 separate top-level strings.

**Comment by mukobi** [comment]
Addressed Evan's changes, ready for more review. It was a bit ugly merging in changes from #36 as we both used slightly different ways of doing things in pageserve/test_routes.py, but I think it's good now and it passes all the tests.

File: pageserve/test_routes.py [hunk 12]
```diff
@@ -1,39 +1,154 @@
...
+    def test_valid_authentication(self, requests_mocker):
+        """GAuth token successfully authenticates."""
+        with app.app.test_request_context(), app.app.test_client() as client:
+            requests_mocker.post(app.app.config["USERS_ENDPOINT"] + "authenticate",
+                                 json=VALID_SESSION,
+                                 status_code=VALID_USER_AUTH_STATUS)
+            result = client.post("/v1/authenticate", data={
+                "gauth_token": "I don't matter because requests is mocked"})
+            self.assertEqual(result.status_code, VALID_USER_AUTH_STATUS)
+            result_dict = ast.literal_eval(result.data.decode())
```
Comment (cmei4444) [inline pageserve/test_routes.py]: Why do you need ast here when you're just decoding a dict? Especially since this is the only place you use it.

Comment (mukobi) [reply pageserve/test_routes.py]: JSON using double quotes for strings. `client.post()` returns a Python dict-like string, not JSON, that has single quoted keys, so `json.loads()` raises an exception on it. `ast.literal_eval()` here is used to decode the result into a Python dict correctly.

File: pageserve/app.py [hunk 13]
```diff
@@ -15,47 +15,94 @@
...
 
 
 @app.route('/v1/', methods=['GET'])
 def index():
     """Displays home page with all past posts."""
-    user = get_user()
-    is_auth = has_edit_access(get_user_info(user,
-                                            app.config['USERS_ENDPOINT']
-                                            + "authorization"))
     posts = get_posts()
```
Comment (whaught) [inline pageserve/app.py]: just wondering, why assign posts to a temp variable, but then we inlined 'auth='

File: pageserve/app.py [hunk 14]
```diff
@@ -151,22 +198,29 @@ def parse_events(events_dict):
...
-
-def has_edit_access(user_data):
+def has_edit_access(user):
     """Determines if the user with the given info has edit access."""
-    return user_data['edit_access']
+    if user is None:
+        return False
+    url = app.config["USERS_ENDPOINT"] + "authorization"
+    response = requests.post(url, data={"user_id": user["user_id"]})
+    return bool(response.json()["edit_access"]) if user else False
```
Comment (whaught) [inline pageserve/app.py]: what's the value of 'edit_access'? do we want to do a comparison rather than cast to bool?

Comment (mukobi) [reply pageserve/app.py]: `response.json()` confusingly returns a Python dict, so `["edit_access"]` will be a bool. The `if user` check is redundant from above and casting to bool might be a little unsafe, so I just compared to `True` in https://github.com/knative-portability/large-events/pull/37/commits/0ff93b6261cef48ed24ed0082ab6793027e6031a

File: pageserve/app.py (hunk 13 above)
Comment (mukobi) [reply pageserve/app.py]: Thanks, not sure why, it was like that when I found it. This and the same things with `events = get_events` fixed in https://github.com/knative-portability/large-events/pull/37/commits/0ff93b6261cef48ed24ed0082ab6793027e6031a

**Comment by mukobi** [comment]
Addressed reviews, ready for more.

**Pull request merged**